### ISSUE TITLE
Bundler support

### ISSUE THREAD

**Issue by damien** [misogi/vscode-ruby-rubocop#34]
Hey @misogi!

First off, thanks for writing this extension! Super useful!

I'd like to contribute to this repository, I was thinking of adding support for detecting projects that have a Gemfile/Gemfile.lock and using that to invoke rubocop via `bundle exec`. I think this may also solve the open issue of support via rbenv (Bundler should handle setting up the environment for the executable).

Would this be a welcome contribution? Thoughts?

**Comment by misogi** [comment]
Hi @damien !

Thanks for your suggestion.
Your contribution about bundler is welcome.

I'm also using bundler. This is a good idea!
If PATH is correct. Extention can execute bundler like rubocop.
About PATH issue. A problem of searching bundler itself will remain. we solve this individualy.

Detecting Gemfile is nice. but when if Gemfile has no rubocop, how handle this case?
I'm grad to configure to switch using bundler.
(vscode-ruby has useBundler option.)

**Comment by damien** [comment]
Thanks for the quick response @misogi!

> If PATH is correct. Extention can execute bundler like rubocop. About PATH issue. A problem of searching bundler itself will remain. we solve this individualy.

I was thinking about this. I imagine the big barrier to support for tools like rbenv would be ensuring that path is set with the correct context—possibly by invoking `rbenv init -` or by trying to detect rbenv configuration variables in the environment variables visible to VSCode.

> Detecting Gemfile is nice. but when if Gemfile has no rubocop, how handle this case?

Good question! I was thinking of lookup/fallback mechanism for determining how to find/select a Rubocop executable to use. As a rough proposal, something like this:

1. Determine if the currently open VSCode project/folder is using Bundler
2. If Bundler is detected, determine if Rubocop is present in the Bundler dependencies.
3. If Rubocop is present in the bundled dependencies, use it.
4. If Rubocop is _not_ present within the Bundled dependencies, attempt to find it on the user's PATH (the current behavior/implementation)

> I'm grad to configure to switch using bundler.
(vscode-ruby has useBundler option.)

This is an interesting thought! I'll dig into `vscode-ruby` and see how that extension handles Bundler detection. It may be that this is a solved problem and we can build on a similar executable detection/invocation mechanism. :)

<hr>

So, to summarize, the open questions I need to address are as follows:

1. How will we detect rbenv in the current PATH or environment?
2. How will we detect Bundler in the current PATH or environment?
3. What should the overall executable lookup/detection mechanism be? (eg. If we detect Rubocop present in Bundler and elsewhere on the user's PATH, which do we prefer?)

I'm also starting to think these may be a couple separate units of work. I'm fine making a PR to add either rbenv or Bundler support, do you have any preference on which should be done first?

Let me know if you have any other thoughts or questions!

**Comment by misogi** [comment]
Thanks! I'm happy you sort out the issue. :+1: 

I want you to PR separately.
I think add feature to support bundler first. And then It will support rbenv.

> 1 How will we detect rbenv in the current PATH or environment?

That make sense. I understand extension detect installation of rbenv itself. right?
Invoking `rbenv init` on node or using  `$RBENV_ROOT` .. There are likely solution to solve PATH.

I have been troubled about loading path of external Ruby management tool. (rbenv and others. long time :( )
This approach will may discover a hint of detect path in like rvm.

> 2 How will we detect Bundler in the current PATH or environment?

In my opinion, this is most useful feature for many Ruby developer.
Detect Gemfile and execute `bundle exec` instead of rubocop on current PATH detection logic. and then other PR solve loading rbenv PATH.

> 3 What should the overall executable lookup/detection mechanism be?

Your proposal for detection rubocop is great. I think extension choose bundler's rubocop first too.
If this is possible, we may don't have to configure like "useBundler".

**Comment by damien** [comment]
Thanks again for the feedback @misogi!

It sounds like we want to start preferring convention to configurations like `ruby.rubocop.executePath` and `ruby.rubocop.configFilePath`. Based on that and the discussion we've had so far, I'll get started on adding Bundler support.

To be clear, I interpret "bundler support" to mean that we check for Bundler and the presence of Rubocop in the gem bundle. If Rubocop is present in the bundle, we use that version of Rubocop. Otherwise we fall back to current behavior.

I'll also probably be stubbing out a basic test suite for the current functionality of this extension. I'll feel a lot more confident adding new things if we can verify that I haven't added any new bugs. Do you have any preference for a specific Javascript unit testing library? I'll probably end up using Jasmine or Mocha unless told otherwise.

**Comment by misogi** [comment]
Hi @damien .
Thanks for the explanation.

Sorry. Let's get back to talk about "bundler support".
I think falling back to current behavior is our goal too.

I prefer to test TDD style. I would choice Mocha/Chai(assert) style. Do you know any cool/modern choice?

**Comment by damien** [comment]
Mocha works fine for me. Never used Chai, but I'll pull it in and give it a shot! Reminds me a lot of Shoulda.

TDD is also fine by me.

**Comment by damien** [comment]
As for Bundler support: I agree. Any new logic added to support Bundler should fall back to the current implementation when Bundler is not available or does not have Rubocop in the Gemfile.

**Comment by misogi** [comment]
@damien If testing with chai is hard. Current project's package has Mocha/(Node assert) style. It is enough good.
